### ISSUE TITLE
QTable: Group keys don't have a unit

### ISSUE THREAD

**Issue by olebole** [astropy/astropy#14964]
### Description

When grouping a QTable with a column that contains a quantity, the keys only contain a float with a value, but not the full quantity:

```Python
>>> from astropy.table import QTable
>>> import astropy.units as u
>>> tbl = QTable({"length": [1., 1., 2., 3., 1., 2.,]*u.m})
>>> gtbl = tbl.group_by('length')
>>> for cols in gtbl.groups.keys:
...     print(cols)
...     print(dict(cols))
... 
length
  m   
------
   1.0
{'length': 1.0}
length
  m   
------
   2.0
{'length': 2.0}
length
  m   
------
   3.0
{'length': 3.0}
```

### Expected behavior

The keys should be complete, i.e. for a quantity column it should be a quantity with the proper unit.

### Versions

* Linux-6.1.0-9-amd64-x86_64-with-glibc2.36 (Debian bookworm)
* Python 3.11.4 (main, Jun  7 2023, 10:13:09) [GCC 12.2.0]
* astropy 5.2.1 (also checked with 5.3)
* Numpy 1.24.2
* pyerfa 2.0.0.3
* Scipy 1.10.1
* Matplotlib 3.6.3

**Comment by taldcroft** [comment]
@olebole - this should be possible without too much disruption. I have an idea for a lightweight fix, but it might also be done as part of #14942.

**Comment by taldcroft** [comment]
Potential fix in #14966.

**Comment by olebole** [comment]
Thanks for the quick fix! I also found a workaround for the time being: wrap the keys into a **QTable**:
```Python
>>> …
>>> for cols in QTable(gtbl.groups.keys):
...     print(dict(cols))
{'length': <Quantity 1. m>}
{'length': <Quantity 2. m>}
{'length': <Quantity 3. m>}
```
